stout/iterate.h

```diff
@@ -0,0 +1,136 @@
+#include <array>
+#include <iterator>
+
+#include "stout/context.h"
+#include "stout/loop.h"
+#include "stout/stream.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace stout {
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename Iter>
+auto Iterate(Iter begin, Iter end) {
+  struct ContainerData {
+    Iter it_;
+    Iter end_;
+  };
+
+  using T = typename std::iterator_traits<Iter>::value_type;
+
+  return Stream<T>()
+      .context(ContainerData{begin, end})
+      .next([](auto& containerData, auto& k) {
+        if (containerData.it_ != containerData.end_) {
+          k.Emit(*(containerData.it_++));
+        } else {
+          k.Ended();
+        }
+      })
+      .done([](auto&, auto& k) {
+        k.Ended();
+      });
```

Comment:
Whenever we only need scope to be captured in a single lambda let's just use lambda captures directly. So in this case:

```suggestion
  return Stream<T>()
      .next([begin, end](auto& k) mutable {
        if (begin != end) {
          k.Emit(*(begin++));
        } else {
          k.Ended();
        }
      })
      .done([](auto& k) {
        k.Ended();
      });
```

Please update here and the rest of the review.